File: week1_programming_challenges_starters/2_maximum_pairwise_product/max_pairwise_product.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

// 2.1 Naive Algorithm
func MaxPairwiseProductNaive(dataset []int) int {
	var product int
	n := len(dataset)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			//fmt.Println(dataset[i], dataset[j])
			product = int(math.Max(float64(product), float64(dataset[i])*float64(dataset[j])))
		}
	}
	return product
}

//  2.2 Fast algorithm

func MaxPairwiseProductFast(dataset []int) int {
	index1 := 0
	var index2 int
	n := len(dataset)
	for i := 1; i < n; i++ {
		if dataset[i] > dataset[index1] {
			index1 = i
		}
	}
	if index1 == 0 {
		index2 = 1
	} else {
		index2 = 0
	}

	for i := 0; i < n; i++ {
		if i != index1 && dataset[i] > dataset[index2] {
			index2 = i
		}
	}
	//debuggin fmt.Println(index1, index2)
	return dataset[index1] * dataset[index2]
	// write data set
	//func writedata(

}

// reads first line and second line of input the produces data set
func getDataset(n int) []int {
	nRawSeq := bufio.NewReader(os.Stdin) // next line input: A1,...,An
	elems, _, _ := nRawSeq.ReadLine()
	rawElems := strings.Split(string(elems), " ") //
	cleanElems := make([]int, n)

	for i, num := range rawElems {
		cleanElems[i], _ = strconv.Atoi(num)
	}
	return cleanElems
}

func main() {
	//  Constraints 2<= n <= 2*10exp5 and A1,...An <= 2*10exp5
	var n int
	fmt.Scanf("%d", &n) // first line input: n
	sequence := getDataset(n)
	fmt.Println(MaxPairwiseProductFast(sequence))

	// STRESS TEST
	/*
		var int m
		fmt.Scanf("%d", &m
		StressTest(n, m)

	*/
}

File: week1_programming_challenges_starters/2_maximum_pairwise_product/max_pairwise_product.py
# Uses python3
"""
n = int(input())
a = [int(x) for x in input().split()]
assert(len(a) == n)

"""

def max_pairwise_product(n, a):
    """ Naive but correct algorithm"""
    for i in range(0, n):
        for j in range(i + 1, n):
            if a[i] * a[j] > result:
                result = a[i] * a[j]
    return result


def max_pairwise_product_fast(n, an):
    """Fast implementation"""
    index1 = 0 
    for i in range(1,n):
        if an[i] > an[index1]:
            index1 = i
    if index1 == 0:
        index2 =  1
    else:
        index2 = 0

    for i in range(n):
        if i != index1 and an[i] > an[index2]:
            index2 = i

    return an[index1] * an[index2]


if __name__ == '__main__':
    try:
        n = int(input())
        a = [int(x) for x in input().split()]
    except EOFError:
        pass
    assert (len(a) == n)

print(max_pairwise_product_fast(n, a))
